### PR TITLE
Refactoring.

### PR DESCRIPTION
**Pull request by duckinator**
I made an attempt at refactoring this. Let me know what you think!

### CODE REVIEW

**Comment by donaldpipowitch** [comment]
Hi! Thank you so much for the pull request. I think I'll need some time to look at everything. But I already see some things I really like:
- `fn move_cursor`/`fn key`/and so on
- `Key::Char('\n') if state.value.is_empty() => State::done(state),` - I totally forgot that I can add an `if` into the `match` here 😍sooooo much nicer

Can you explain why you explicitly went for an immutable behavior for `fn move_cursor` and so on? Do you "use" it somewhere? (E.g. by comparing and old state vs. a new state or something like that.) I know immutable behavior is "better" most of the time, but in this case I wonder, if it is "needed" or makes the API nicer? 🤔 Has it performance implications in Rust to construct the struct everytime you want to made a change? The mutability would probably be nicely handled by the borrowing already.

Besides that the program doesn't seem to run correctly locally:
- When I try to press "a" and than backspace, my "a" isn't removed.
- When I press "a", space and "a" I get `  aa` instead of `a a`.
- When I press "a" and enter I get `a\nStart typing...` instead of `a\nA\na\nStart typing...`
- When I press "a" and Ctrl+l the uppercased and lowercased value are removed and I only see the line where I entered "a".
- When I press "a", "a" and the left arrows I seem to never get to the first character.

And at the end we'll need to update the `README`. Maybe we could create smaller pull requests to make "one change" at a time?

**Comment by duckinator** [comment]
hi!

# Immutable Behavior

With regards to choosing immutable behavior, I'm going to be a bit verbose because I find this kind of thing interesting. I included a tl;dr at the end of this section, however. :slightly_smiling_face: 

The essay [Values and Change: Clojure’s approach to Identity and State](https://clojure.org/about/state) explains it well (even if it's about Clojure, not Rust):

> We need to move away from a notion of state as "the content of this memory block" to one of "the value currently associated with this identity". Thus **an identity can be in different states at different times, but the state itself doesn’t change.** That is, an identity is not a state, an identity *has* a state. Exactly one state at any point in time. And that state is a true value, i.e. it never changes. **If an identity appears to change, it is because it becomes associated with different state values over time.** This is the Clojure model.

> In Clojure’s model, value calculation is purely functional. Values never change. **New values are functions of old, not mutations.**

That is the approach I tried to take. 

I feel it works well for this, because instead of having a collection of functions that _modify_ state and update the representation you effectively have:

1. a **state machine** which calculates the next state in a deterministic way.
2. a **renderer** which takes a given state, and creates a rendition of it.
3. a slightly-obfuscated <pre>event processor</pre> which, each time it receives an event (keyboard input):
    * passes that event and the _last known state_ to the _state machine_ to determine the next state,
    * replaces the _last known state_ with the new state,
    * passes the new state to the _renderer_ to pass this information to the user. 

By "slightly-obfuscated", I simply mean it's hidden in a `for`/`in` loop instead of being represented in a way I'd immediately associate with an event system, but not hidden enough to be _completely_ misleading.

**tl;dr:** By separating the acts of **calculating the next state** and **rendering**, you can create a chain of <code>(last state, event) -&gt; state calculation -&gt; rendering</code>, while **preserving the ability to reason about each component in isolation**.

Whether or not that's _useful_ or just interesting-to-me theory, I'm not so sure. :slightly_smiling_face:

**Comment by duckinator** [comment]
hi!

# Immutable Behavior

With regards to choosing immutable behavior, I'm going to be a bit verbose because I find this kind of thing interesting. I included a tl;dr at the end of this section, however. :slightly_smiling_face: 

The essay [Values and Change: Clojure’s approach to Identity and State](https://clojure.org/about/state) explains it well (even if it's about Clojure, not Rust):

> We need to move away from a notion of state as "the content of this memory block" to one of "the value currently associated with this identity". Thus **an identity can be in different states at different times, but the state itself doesn’t change.** That is, an identity is not a state, an identity *has* a state. Exactly one state at any point in time. And that state is a true value, i.e. it never changes. **If an identity appears to change, it is because it becomes associated with different state values over time.** This is the Clojure model.

> In Clojure’s model, value calculation is purely functional. Values never change. **New values are functions of old, not mutations.**

That is the approach I tried to take. 

I feel it works well for this, because instead of having a collection of functions that _modify_ state and update the representation at the same time, you effectively have:

1. a **state machine** which calculates the next state in a deterministic way.
2. a **renderer** which takes a given state, and creates a rendition of it.
3. a slightly-obfuscated **event processor** which, each time it receives an event (keyboard input):
    * passes that event and the _last known state_ to the _state machine_ to determine the next state,
    * replaces the _last known state_ with the new state,
    * passes the new state to the _renderer_ to pass this information to the user. 

By "slightly-obfuscated", I simply mean it's hidden in a `for`/`in` loop instead of being represented in a way I'd immediately associate with an event system, but not hidden enough to be _completely_ misleading.

**tl;dr:** By separating the acts of **calculating the next state** and **rendering**, you can create a chain of <code>(last state, event) -&gt; state calculation -&gt; rendering</code>, while **preserving the ability to reason about each component in isolation**.

Whether that's _useful_ or just interesting-to-me theory, I'm not so sure. :slightly_smiling_face: 

## Bugs

With regard to the bugs, the only one I can reproduce locally (Fedora 28, using the urxvt terminal) is the last one. When I have some free time (this weekend or next week), I'll try it out in different terminals and see if I can fix that and reproduce the others.

## `README`/smaller PRs

That seems reasonable to me. Should I use this PR as basically a prototyping phase, then, to figure out an overall approach?

**Comment by donaldpipowitch** [comment]
> That seems reasonable to me. Should I use this PR as basically a prototyping phase, then, to figure out an overall approach?

Sorry for the late response. Yeah, some small PRs would be nice. You could start with some of the updated comments and then with `Key::Char('\n') if state.value.is_empty()` for example. If you don't have time to do that I could also add this based on your current PR here.
